File: problems/2095/Lc2095.ts
/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
  static fromArray(arr: number[]): ListNode | null {
    if (arr.length) {
      const head = new ListNode(arr[0]);
      let node: ListNode = head;
      for (let i = 1; i < arr.length; i += 1) {
        node.next = new ListNode(arr[i]);
        node = node.next;
      }
      return head;
    }
    return null;
  }
  toArray(): number[] {
    const result: number[] = [];
    let node: ListNode | null = this;
    while (node != null) {
      result.push(node.val);
      node = node.next;
    }
    return result;
  }
}

function deleteMiddle(head: ListNode | null): ListNode | null {
  if (head?.next == null) return null;
  let bnode: ListNode | null = head,
    fnode: ListNode | null = head.next.next;
  while (fnode != null) {
    fnode = fnode.next;
    if (fnode != null && bnode != null) {
      bnode = bnode.next;
      fnode = fnode.next;
    }
  }
  if (bnode?.next != null)
    bnode.next = bnode.next.next;
  return head;
};

function arraysEqual(arr1: any[], arr2: any[]) {
  return (arr1.length === arr2.length
    && arr1.every((val, pos) => (val === arr2[pos])));
};

function main() {
  let head: ListNode | null, node: ListNode | null, result: number[], expected: number[];

  head = ListNode.fromArray([1, 3, 4, 7, 1, 2, 6]);
  node = deleteMiddle(head);
  result = (node == null ? [] : node.toArray());
  expected = [1, 3, 4, 1, 2, 6];
  if (!arraysEqual(expected, result)) new Error(`expected ${expected}, but ${result}`);

  head = ListNode.fromArray([1, 3, 4, 7, 1, 2]);
  node = deleteMiddle(head);
  result = (node == null ? [] : node.toArray());
  expected = [1, 3, 4, 1, 2];
  if (!arraysEqual(expected, result)) new Error(`expected ${expected}, but ${result}`);
};
main();

File: problems/2095/Lc2095.py
from typing import Optional
from typing import List


class ListNode:
    """Definition for singly-linked list."""

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    @classmethod
    def fromList(cls, values=()):
        head = None
        node = None
        for pos, val in enumerate(values):
            if 0 < pos:
                node.next = cls(val)
                node = node.next
            else:
                head = node = cls(val)
        return head

    def toList(self) -> List[int]:
        result = []
        node = self
        while node:
            result.append(node.val)
            node = node.next
        return result


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head.next:
            return None
        bnode, fnode = head, head.next.next  # head -> bnode -> fnode
        while fnode:
            fnode = fnode.next
            if fnode:
                bnode, fnode = bnode.next, fnode.next
        bnode.next = bnode.next.next
        return head


def main():
    sol = Solution()

    head = ListNode.fromList([1, 3, 4, 7, 1, 2, 6])
    result = sol.deleteMiddle(head).toList()
    expected = [1, 3, 4, 1, 2, 6]
    assert expected == result, f'expected {expected}, but {
        result}'

    head = ListNode.fromList([1, 3, 4, 7, 1, 2])
    result = sol.deleteMiddle(head).toList()
    expected = [1, 3, 4, 1, 2]
    assert expected == result, f'expected {expected}, but {
        result}'


if __name__ == '__main__':
    main()
